### PR TITLE
Implement "Mixed Content" check

### PR DESCRIPTION
**Pull request by jugglinmike**
Commit message:

> This check ensures that resources embedded within HTML documents are not
> requested using a protocol that degrades the security of the original
> connection. This avoids "mixed content" warnings in browsers that
> enforce consistent protocols, and it precludes the leaking of private
> information in browsers that do not.

Resolves gh-542

### CODE REVIEW

File: lib/nanoc/extra/link_collector.rb [hunk 1]
```diff
@@ -4,6 +4,18 @@
 
 module ::Nanoc::Extra
   class LinkCollector
+
+    @@uri_attrs = {
```
Comment (ddfreyne) [inline lib/nanoc/extra/link_collector.rb]: I’d prefer to have this in a constant rather than a class variable (I think Rubocop will complain anyway).

**Comment by ddfreyne** [comment]
This looks good! I’ll do a more in-depth review soon, but I‘m quite liking it so far.

**Comment by jugglinmike** [comment]
Great, thanks! I'll hold off on moving that class variable till you get a chance to finish.

File: lib/nanoc/extra/link_collector.rb [hunk 2]
```diff
@@ -4,6 +4,18 @@
...
+      'a' => :href,
```
Comment (ddfreyne) [inline lib/nanoc/extra/link_collector.rb]: Does it make sense to disallow linking to a non-secure site from a secure site?

File: lib/nanoc/extra/link_collector.rb [hunk 3]
```diff
@@ -33,24 +45,46 @@ def filenames_per_href
...
+      filenames_per_resource_uri
+    end
+
     def external_href?(href)
       href =~ %r{^(\/\/|[a-z\-]+:)}
     end
 
-    def hrefs_in_file(filename)
-      hrefs_in_file = Set.new
+    def uris_in_file(filename, tag_names)
```
Comment (ddfreyne) [inline lib/nanoc/extra/link_collector.rb]: Renaming this method breaks backwards compatibility of the API. Could you change it back, or let `hrefs_in_file` proxy to the new method?

(I’m not sure whether `LinkChecker` should’ve been part of the public API, but it’s too late now anyway!)

Comment (ddfreyne) [inline lib/nanoc/extra/link_collector.rb]: Nevermind, just saw that you already did that. Sweet :)

Comment (ddfreyne) [inline lib/nanoc/extra/link_collector.rb]: (FYI: you can mark classes/methods/modules/… as not part of the public API using `@api private`.)

File: lib/nanoc/extra/checking/checks/mixed_content.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        next unless guaranteed_insecure?(uri)
+        fns.each do |filename|
+          add_issue(
+            "mixed content include: #{uri}",
+            subject: filename)
+        end
+      end
+    end
+
+    protected
```
Comment (ddfreyne) [inline lib/nanoc/extra/checking/checks/mixed_content.rb]: I go for `private` only these days as I find `protected` to be not so useful.

File: lib/nanoc/extra/checking/checks/mixed_content.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    @@protocol_pattern = /^(\w+):\/\//
```
Comment (ddfreyne) [inline lib/nanoc/extra/checking/checks/mixed_content.rb]: Same comment about the constant vs class variable as below.

Comment (ddfreyne) [inline lib/nanoc/extra/checking/checks/mixed_content.rb]: (I think you might have to use `@private` to hide it in the YARD documentation.)

**Comment by ddfreyne** [comment]
The style checks are failing. Apparently, they’re also failing on other parts of the code—I’ll fix it; ignore the style test results for now.

File: lib/nanoc/extra/link_collector.rb (hunk 2 above)
Comment (jugglinmike) [inline lib/nanoc/extra/link_collector.rb]: 1. Prevent insertion of insecure "active content" (such as JavaScript) that could generate misleading content or steal private information
2. Prevent passive leaking of information--HTTP requests are sent in-the-clear and might leak information about the user's browsing session

Users who navigate using insecure hyperlinks are susceptible to `#2`, but navigation is performed actively by the user, and it's harder to achieve the task of "saving users from themselves". Plus, the current site's maintainer has no control over the availability of SSL on remote servers.

**Comment by ddfreyne** [comment]
Can you rebase onto master? That should fix the style issues (at least the ones that are not your fault, heh).

**Comment by ddfreyne** [comment]
Looks good otherwise!

**Comment by jugglinmike** [comment]
Alrighty, I've rebased and incorporated your feedback in separate commits. I'm happy to squash when you think this is ready

File: lib/nanoc/extra/checking/checks/mixed_content.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# encoding: utf-8
+
+PROTOCOL_PATTERN = /^(\w+):\/\//
```
Comment (ddfreyne) [inline lib/nanoc/extra/checking/checks/mixed_content.rb]: Could you move this into the `MixedContent` class? This way, `PROTOCOL_PATTERN` won’t be a top-level constant. Constant lookup will still work.

File: lib/nanoc/extra/link_collector.rb [hunk 7]
```diff
@@ -2,6 +2,17 @@
 
 require 'set'
 
+URI_ATTRS = {
```
Comment (ddfreyne) [inline lib/nanoc/extra/link_collector.rb]: Same as above (move into LinkCollector).

File: lib/nanoc/extra/checking/checks/mixed_content.rb (hunk 6 above)
Comment (jugglinmike) [inline lib/nanoc/extra/checking/checks/mixed_content.rb]: Ah yeah, much better that way.

**Comment by ddfreyne** [comment]
Cool! OK to squash. I’ll properly test it out tomorrow and will merge it if it’s OK!

**Comment by jugglinmike** [comment]
Excellent--all squashed

**Comment by ddfreyne** [comment]
Thanks!

**Comment by jugglinmike** [comment]
My pleasure :)

**Pull request merged**